ray-operator/controllers/ray/common/service.go

```diff
@@ -40,27 +41,72 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
 		annotations = make(map[string]string)
 	}
 
+	default_name := utils.GenerateServiceName(cluster.Name)
+	default_namespace := cluster.Namespace
+	default_type := cluster.Spec.HeadGroupSpec.ServiceType
+
+	defaultAppProtocol := DefaultServiceAppProtocol
+	ports_int := getServicePorts(cluster)
+	ports := []corev1.ServicePort{}
+	for name, port := range ports_int {
+		svcPort := corev1.ServicePort{Name: name, Port: port, AppProtocol: &defaultAppProtocol}
+		ports = append(ports, svcPort)
+	}
+	if cluster.Spec.HeadGroupSpec.HeadService != nil {
+		// Use the provided "custom" HeadService.
+		// Deep copy the HeadService to avoid modifying the original object
+		headService := cluster.Spec.HeadGroupSpec.HeadService.DeepCopy()
+
+		// Merge labels with custom HeadService labels. If there are overlaps,
+		// ignore the custom HeadService labels.
+		for k, v := range labels {
+			headService.ObjectMeta.Labels[k] = v
+		}
+
+		// For the selector, use labels ignore any custom HeadService selectors or labels.
+		headService.Spec.Selector = labels
+
+		// Merge annotations with custom HeadService annotations. If there are overlaps,
+		// ignore the custom HeadService annotations.
+		for k, v := range annotations {
+			headService.ObjectMeta.Annotations[k] = v
+		}
+
+		// Append default ports.
+		headService.Spec.Ports = append(headService.Spec.Ports, ports...)
+
+		// If the user has not specified a name, generate one
+		if headService.ObjectMeta.Name == "" {
+			headService.ObjectMeta.Name = default_name
+		}
+
+		// If the user has not specified a namespace, use the cluster's namespace
+		if headService.ObjectMeta.Namespace == "" {
+			headService.ObjectMeta.Namespace = default_namespace
+		}
+
+		// If the user has not specified a service type, use the cluster's service type
+		if headService.Spec.Type == "" {
+			headService.Spec.Type = default_type
+		}
+
+		return headService, nil
+	}
+
 	service := &corev1.Service{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:        utils.GenerateServiceName(cluster.Name),
-			Namespace:   cluster.Namespace,
+			Name:        default_name,
+			Namespace:   default_namespace,
 			Labels:      labels,
```

Comment:
`Labels` and `Selector` are the same here. Will this cause connection issues for worker/head Pods if users update the value of custom labels for the head service?

Reply:
Discussed offline. Conclusion for Selector: Only use the keys from (2), don't add any new keys. Just update the values of the keys from (2) with the values from (1) if (1) has those keys. Always ignore keys and values from (3))